gr-soapy/lib/block_impl.cc

```diff
@@ -0,0 +1,664 @@
+
+namespace gr {
+namespace soapy {
+
+const pmt::pmt_t CMD_CHAN_KEY = pmt::mp("chan");
+const pmt::pmt_t CMD_FREQ_KEY = pmt::mp("freq");
+const pmt::pmt_t CMD_GAIN_KEY = pmt::mp("gain");
+const pmt::pmt_t CMD_ANTENNA_KEY = pmt::mp("ant");
+const pmt::pmt_t CMD_RATE_KEY = pmt::mp("rate");
+const pmt::pmt_t CMD_BW_KEY = pmt::mp("bw");
+
+/*
+ * The private constructor
+ */
+block_impl::block_impl(int direction,
+                       const std::string& device,
+                       const std::string& type,
+                       size_t nchan,
+                       const std::string& dev_args,
+                       const std::string& stream_args,
+                       const std::vector<std::string>& tune_args,
+                       const std::vector<std::string>& other_settings)
+    : d_direction(direction),
+      d_nchan(nchan),
+      d_stream_args(stream_args),
+      d_channels(nchan)
+{
+    // Channel list
+    std::iota(std::begin(d_channels), std::end(d_channels), 0);
+
+    // Must be 1 or nchan.
+    if (nchan != tune_args.size() && tune_args.size() != 1) {
+        throw std::invalid_argument(name() +
+                                    ": Wrong number of channels and tune arguments");
+    }
+
+    // Must be 1 or nchan.
+    if (nchan != other_settings.size() && other_settings.size() != 1) {
+        throw std::invalid_argument(name() + ": Wrong number of channels and settings");
+    }
+
+    // Convert type to Soapy type.
+    if (type == "fc32") {
+        d_soapy_type = SOAPY_SDR_CF32;
+    } else if (type == "sc16") {
+        d_soapy_type = SOAPY_SDR_CS16;
+    } else if (type == "sc8") {
+        d_soapy_type = SOAPY_SDR_CS8;
+    } else {
+        throw std::invalid_argument(name() + ": Invalid IO type");
+    }
+
+    SoapySDR::Kwargs dev_kwargs = SoapySDR::KwargsFromString(device + ", " + dev_args);
+    d_device = device_ptr_t(SoapySDR::Device::make(dev_kwargs));
+
+    /* Some of the arguments maybe device settings, so we need to re-apply them */
+    SoapySDR::ArgInfoList supported_settings = d_device->getSettingInfo();
+    for (const auto& arg : dev_kwargs) {
+        for (const auto& supported : supported_settings) {
+            if (supported.key == arg.first) {
+                d_device->writeSetting(arg.first, arg.second);
+            }
+        }
+    }
+
+    if (d_nchan > d_device->getNumChannels(d_direction)) {
+        throw std::invalid_argument(
+            name() + ": Unsupported number of channels. Only  " +
+            std::to_string(d_device->getNumChannels(d_direction)) +
+            " channels available.");
+    }
+
+    /*
+     * Validate stream arguments for all channels
+     */
+    for (const auto channel : d_channels) {
+        SoapySDR::ArgInfoList supported_args =
+            d_device->getStreamArgsInfo(d_direction, channel);
+        SoapySDR::Kwargs stream_kwargs = SoapySDR::KwargsFromString(stream_args);
+
+        for (const auto& arg : stream_kwargs) {
```

Comment:
This is a fair chunk of repeated code to check for supported args. Worth breaking out to a function, right?

Reply:
Actually and then again with the `settings_kwargs`, looks like.

Reply:
static helper function added ... we could probably get this file down to one line eventually :-)